File: app/src/main/java/com/inelasticcollision/recipelink/dialogs/imagepicker/ImagePickerAdapter.java
/*
 * ImagePickerAdapter.java
 * RecipeLink-SQLBrite
 *
 * Created by Tyler Madonna on 7/15/2017.
 * Copyright © 2017 Tyler Madonna. All rights reserved.
 */

package com.inelasticcollision.recipelink.dialogs.imagepicker;

import android.content.Context;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.inelasticcollision.recipelink.R;

import java.util.List;

class ImagePickerAdapter extends RecyclerView.Adapter<ImagePickerAdapter.ImageViewHolder>{

    private final Context mContext;

    private final List<String> mImages;

    private final ImagePickerAdapter.OnImagePickerAdapterItemClickListener mListener;

    ImagePickerAdapter(Context context, List<String> images, OnImagePickerAdapterItemClickListener listener) {

        mContext = context;

        mImages = images;

        mListener = listener;

    }

    @Override
    public ImageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ImagePickerAdapter.ImageViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_image_picker, parent, false));
    }

    @Override
    public void onBindViewHolder(ImageViewHolder holder, int position) {

        String image = mImages.get(position);

        // Load recipe image
        Glide.with(mContext)
                .load(image)
                .centerCrop()
                .dontAnimate()
                .dontTransform()
                .crossFade()
                .into(holder.image);

    }

    @Override
    public int getItemCount() {
        return mImages != null ? mImages.size() : 0;
    }

    interface OnImagePickerAdapterItemClickListener {

        void onImagePickerAdapterItemClick(String image);

    }

    class ImageViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        ImageView image;

        ImageViewHolder(View itemView) {
            super(itemView);
            image = (ImageView) itemView;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {

            String imageUrl = mImages.get(getAdapterPosition());

            mListener.onImagePickerAdapterItemClick(imageUrl);

        }

    }

}

File: app/src/main/java/com/inelasticcollision/recipelink/ui/fragment/imagepicker/ImagePickerAdapter.kt
package com.inelasticcollision.recipelink.ui.fragment.imagepicker

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import coil.load
import com.inelasticcollision.recipelink.databinding.ItemImagePickerBinding

class ImagePickerAdapter : RecyclerView.Adapter<ImagePickerAdapter.ViewHolder>() {

    private var imageUrls: List<String> = emptyList()

    fun submitImageUrls(imageUrls: List<String>) {
        this.imageUrls = imageUrls
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            ItemImagePickerBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindRecipe(imageUrls[position])
    }

    override fun getItemCount(): Int {
        return imageUrls.size
    }

    // ViewHolder
    class ViewHolder(
        private val binding: ItemImagePickerBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bindRecipe(imageUrl: String?) {
            // Load image
            binding.imageView.load(imageUrl) {
                crossfade(true)
            }

            binding.imageView.setOnClickListener { view ->
                val handle = view.findNavController().previousBackStackEntry?.savedStateHandle
                handle?.set(ImagePickerFragment.IMAGE_PICKER_SAVED_STATE, imageUrl)
                view.findNavController().popBackStack()
            }
        }
    }
}